File: ink_engine/core/storage.py
"""通用存储服务（Kilo Code 式：引擎内全部持久化统一走存储服务接口）。

structured records + JSON 数据，适配内存/sqlite/postgres，连接串/配置
切换后端。引擎承载两类持久化：
- checkpoint 版本链（快照，乐观锁并发写保护）；
- 执行事件日志（append-only，恢复 = 快照 + 增量日志重放）。

接口只定义存储语义，不绑定图执行——未来回合记录/步骤序列/记忆/世界状态
同走此接口（structured records 通道）。后端切换 = 换连接串（create_storage
工厂），业务代码零改动。
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from .events import EngineEvent
from .logging import get_logger
from .patch_chain import PatchChain
from .security import is_sensitive_key

# 连接串协议前缀 → 后端工厂（memory:// / sqlite:///path / postgresql://）
SCHEME_MEMORY = "memory"
SCHEME_SQLITE = "sqlite"
SCHEME_POSTGRES = "postgresql"

# 补丁链序列化标记键（checkpoint JSON 列内联结构，from_dict 精确还原）
_PATCH_CHAIN_MARKER = "__patch_chain__"
# 引擎消息/工具调用序列化标记键（状态通道可能持有 Message/ToolCall 对象，
# checkpoint 落库前须转为可 JSON 结构，恢复时精确还原）
_MESSAGE_MARKER = "__engine_message__"
_TOOL_CALL_MARKER = "__engine_tool_call__"

logger = get_logger(__name__)


def _jsonable_strip(value: Any) -> Any:
    """状态值 → JSON 可序列化（PatchChain/Message/ToolCall 内联为标记 dict），
    单次递归内联敏感键剥离。

    checkpoint 序列化热路径：剥离与 jsonable 合并为一次遍历（原 strip + jsonable
    两次全量递归）。copy-on-write：子树无敏感键返回原对象。
    """
    if isinstance(value, PatchChain):
        return {
            _PATCH_CHAIN_MARKER: True,
            "base": _jsonable_strip(value.base),
            "patches": [
                {"op": p.op.value, "path": list(p.path), "value": _jsonable_strip(p.value)}
                for p in value.patches
            ],
        }
    from .llm.messages import Message, ToolCall

    if isinstance(value, Message):
        return {_MESSAGE_MARKER: True, "data": _jsonable_strip(value.to_dict())}
    if isinstance(value, ToolCall):
        return {
            _TOOL_CALL_MARKER: True,
            "id": value.id,
            "name": value.name,
            "arguments": value.arguments,
        }
    if isinstance(value, dict):
        result: dict[str, Any] = {}
        changed = False
        for key, item in value.items():
            if is_sensitive_key(key):
                result[key] = ""  # 置空保留（键结构不破坏，下游 .get 恒返回空串）
                changed = True
                continue
            stripped = _jsonable_strip(item)
            if stripped is not item:
                changed = True
            result[key] = stripped
        return result if changed else value
    if isinstance(value, list):
        out = [_jsonable_strip(v) for v in value]
        return out if any(o is not v for o, v in zip(out, value, strict=True)) else value
    if isinstance(value, tuple):
        out = tuple(_jsonable_strip(v) for v in value)
        return out if any(o is not v for o, v in zip(out, value, strict=True)) else value
    return value


def _from_jsonable(value: Any) -> Any:
    """JSON 反序列化还原（标记 dict → PatchChain/Message/ToolCall，其余递归）。"""
    if isinstance(value, dict):
        if value.get(_PATCH_CHAIN_MARKER):
            return PatchChain.from_dict(value)
        if value.get(_MESSAGE_MARKER):
            from .llm.messages import Message

            return Message.from_dict(_from_jsonable(value["data"]))
        if value.get(_TOOL_CALL_MARKER):
            from .llm.messages import ToolCall

            return ToolCall(
                id=value["id"], name=value["name"], arguments=value["arguments"]
            )
        return {k: _from_jsonable(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_from_jsonable(v) for v in value]
    return value


@dataclass(frozen=True, slots=True)
class CheckpointRecord:
    """checkpoint 快照记录（版本链节点）。

    Attributes:
        checkpoint_id: 全局自增 id（版本链锚点）。
        thread_id: 会话/线程 id（版本链归属）。
        node: 恢复锚点（该节点完成后写入的快照，重入从该节点继续）。
        graph_path: 嵌套图路径（恢复定位）。
        state: 通道值快照（api_key 已剥离）。
        parent_id: 版本链父指针（None = 链头）。
        reason: 回合终止原因（reply/止损/超限/异常，None = 未终止）。
        created_at: 写入时间戳（epoch 秒）。
        version: 乐观锁版本号（并发写保护，每次写入 +1）。
        event_seq: 执行事件日志锚点（恢复 = 快照 + 该 seq 之后的增量日志重放）。
        error: 异常快照（reason=error 时携带脱敏后的错误消息，可诊断；None = 无）。
    """

    checkpoint_id: int
    thread_id: str
    node: str | None
    graph_path: tuple[str, ...] = ()
    state: dict = field(default_factory=dict)
    parent_id: int | None = None
    reason: str | None = None
    created_at: float = field(default_factory=time.time)
    version: int = 1
    event_seq: int = 0
    error: str | None = None

    def to_dict(self) -> dict:
        return {
            "checkpoint_id": self.checkpoint_id,
            "thread_id": self.thread_id,
            "node": self.node,
            "graph_path": list(self.graph_path),
            "state": _jsonable_strip(self.state),  # 安全：序列化即剥离敏感键
            "parent_id": self.parent_id,
            "reason": self.reason,
            "created_at": self.created_at,
            "version": self.version,
            "event_seq": self.event_seq,
            "error": self.error,
        }

    @classmethod
    def from_dict(cls, data: dict) -> CheckpointRecord:
        return cls(
            checkpoint_id=int(data["checkpoint_id"]),
            thread_id=data["thread_id"],
            node=data.get("node"),
            graph_path=tuple(data.get("graph_path") or ()),
            state=_from_jsonable(data.get("state") or {}),
            parent_id=data.get("parent_id"),
            reason=data.get("reason"),
            created_at=float(data.get("created_at") or time.time()),
            version=int(data.get("version") or 1),
            event_seq=int(data.get("event_seq") or 0),
            error=data.get("error"),
        )


@runtime_checkable
class Storage(Protocol):
    """通用存储服务接口。

    实现要求：
    - checkpoint 写入支持乐观锁（expected_version 不匹配抛
      CheckpointConflictError，调用方重读后重试）；
    - 事件日志 append-only（seq 严格递增，重放按 seq 有序）；
    - 全部方法幂等/可重试（网络抖动场景调用方重试安全）。
    """

    # ── checkpoint 版本链 ──
    async def get_checkpoint(self, checkpoint_id: int) -> CheckpointRecord | None: ...
    async def get_latest_checkpoint(self, thread_id: str) -> CheckpointRecord | None: ...
    async def put_checkpoint(
        self,
        record: CheckpointRecord,
        *,
        expected_version: int | None = None,
        fork: bool = False,
    ) -> CheckpointRecord: ...
    async def list_checkpoints(self, thread_id: str, *, limit: int = 100) -> list[CheckpointRecord]: ...

    # ── 执行事件日志（append-only）──
    async def append_event(self, thread_id: str, event: EngineEvent) -> int: ...
    async def events_after(self, thread_id: str, seq: int) -> list[EngineEvent]: ...
    async def truncate_events(self, thread_id: str, after_seq: int) -> None: ...
    async def latest_event_seq(self, thread_id: str) -> int: ...

    # ── structured records（回合记录/记忆/世界状态共用）──
    async def put_record(self, collection: str, key: str, data: dict) -> None: ...
    async def get_record(self, collection: str, key: str) -> dict | None: ...
    async def list_records(self, collection: str) -> list[dict]: ...

    async def close(self) -> None: ...


def create_storage(conn_string: str) -> Storage:
    """存储后端工厂：连接串协议前缀决定后端（内存/sqlite/postgres）。

    例：memory://、sqlite:///:memory:、sqlite:///./engine.db、
    postgresql://user:pwd@host/db（postgres:// 别名兼容 DATABASE_URL 形态）
    """
    scheme = conn_string.split(":", 1)[0].lower()
    if scheme == SCHEME_MEMORY or conn_string in ("", "memory"):
        from .storage_memory import MemoryStorage

        return MemoryStorage()
    if scheme == SCHEME_SQLITE:
        from .storage_sqlite import SqliteStorage

        if not conn_string.startswith("sqlite://"):
            # 显式前缀校验：sqlite:/path 等少斜杠形态会静默截断成错误相对路径并新建空库
            raise ValueError(
                f"非法 sqlite 连接串（应为 sqlite:///path 或 sqlite:///:memory:）: {conn_string}"
            )
        # 剥离 "sqlite://" 前缀；:memory: 保留原形（内存库），路径去掉前导 /
        db_path = conn_string[len(SCHEME_SQLITE) + 3 :]
        if not db_path:
            # 空路径归一为内存库（"sqlite" 裸协议默认值），持久化未启用需显式
            logger.warning("sqlite 连接串未指定路径，归一为 :memory:（持久化未启用）")
            db_path = ":memory:"
        if db_path.startswith(":") or db_path.startswith("file:"):
            return SqliteStorage(db_path)
        return SqliteStorage(db_path.lstrip("/"))
    if scheme in (SCHEME_POSTGRES, "postgres"):
        from .storage_postgres import PostgresStorage

        return PostgresStorage(conn_string)
    raise ValueError(f"未知存储连接串协议: {conn_string}")


__all__ = [
    "SCHEME_MEMORY",
    "SCHEME_POSTGRES",
    "SCHEME_SQLITE",
    "CheckpointRecord",
    "Storage",
    "create_storage",
]

File: ink-ts/engine/src/core/storage/storage.ts
/**
 * 存储服务 seam：Storage 接口（checkpoint 版本链 + 事件日志 +
 * structured records + 快照/恢复能力）+ validate_chain 纯校验。
 *
 * seam 仅声明语义：实现由宿主注入（memory/sqlite/postgres 均为宿主侧
 * IO，core 不落）。校验函数消费 seam 的 chain_index + get_checkpoint
 * 两个原语，校验器与后端解耦——可对任意 Storage 实现做存量坏链巡检。
 *
 * validate_chain 镜像 Python 行为：整链一次取回（避免逐跳重查询的
 * O(链长) 次串行 DB 往返）→ 内存内沿 parent_id 回溯 → 遇坏链报告
 * 违规并停止（环检测立即终止，不触发遍历超限）。悬挂/跨线程父指针
 * 单次跨索引查询区分：先看本线程 by_id，缺失再回退 get_checkpoint
 * 判定悬挂或跨线程。
 *
 * 错误映射（与 Python 等价）：值类（参数越界）→ RangeError；类型类
 * → TypeError；其他异常 → new Error。CheckpointConflictError 仍由实现
 * 侧抛（写入端不变量），校验器不涉及。
 */

import type { EngineEvent } from '../events/events.js';

import { ChainLink, CheckpointRecord } from './storage_records.js';
import { DEFAULT_CHAIN_WALK_LIMIT, DEFAULT_LIST_CHECKPOINTS_LIMIT } from './storage_constants.js';

/** 通用存储服务 seam（接口层，宿主侧注入实现）。 */
export interface Storage {
  // ── checkpoint 版本链 ──
  get_checkpoint(checkpoint_id: number): Promise<CheckpointRecord | null>;
  get_latest_checkpoint(thread_id: string): Promise<CheckpointRecord | null>;
  put_checkpoint(
    record: CheckpointRecord,
    opts?: { expected_version?: number | null; fork?: boolean },
  ): Promise<CheckpointRecord>;
  list_checkpoints(
    thread_id: string,
    opts?: { limit?: number },
  ): Promise<CheckpointRecord[]>;
  chain_index(thread_id: string): Promise<ChainLink[]>;
  delete_checkpoints(thread_id: string, ids: readonly number[]): Promise<number>;
  set_checkpoint_parent(
    thread_id: string,
    checkpoint_id: number,
    parent_id: number | null,
  ): Promise<number>;

  // ── 执行事件日志（append-only） ──
  append_event(thread_id: string, event: EngineEvent): Promise<number>;
  events_after(thread_id: string, seq: number): Promise<EngineEvent[]>;
  truncate_events(thread_id: string, after_seq: number): Promise<void>;
  trim_events(thread_id: string, before_seq: number): Promise<number>;
  latest_event_seq(thread_id: string): Promise<number>;

  // ── structured records（回合记录/记忆等宿主结构化数据共用） ──
  put_record(collection: string, key: string, data: Record<string, unknown>): Promise<void>;
  get_record(collection: string, key: string): Promise<Record<string, unknown> | null>;
  list_records(collection: string): Promise<Record<string, unknown>[]>;
  delete_collection(collection: string): Promise<number>;

  // ── 全量快照（备份/迁移/归档） ──
  readonly snapshot_capable: boolean;
  snapshot(dest: string): Promise<void>;
  restore(src: string): Promise<void>;

  close(): Promise<void>;
}

/** validate_chain 选项。 */
export interface ValidateChainOptions {
  /** 回溯步数上限（防意外成环死循环；超限报违规并停止）。 */
  max_walk?: number;
  /** 是否校验 event_seq 单调性（分叉链豁免场景置 false）。 */
  check_event_seq?: boolean;
}

/**
 * 版本链一致性校验：沿 parent_id 回溯，断言父引用存在且归属同线程、
 * checkpoint_id 严格递减、event_seq 单调不减。
 *
 * 实现要点：单次 chain_index 取回整链（轻量行），内存内按 parent_id
 * 回溯。悬挂/跨线程父指针单次跨索引查询区分（看本线程 by_id 是否
 * 命中；不命中回退 get_checkpoint 判定——悬挂或跨线程）。环/自引用
 * 立即终止回溯（继续走无意义且死循环）。
 */
export async function validate_chain(
  storage: Storage,
  thread_id: string,
  options: ValidateChainOptions = {},
): Promise<string[]> {
  const max_walk = options.max_walk ?? DEFAULT_CHAIN_WALK_LIMIT;
  const check_event_seq = options.check_event_seq ?? true;
  const violations: string[] = [];
  const links = await storage.chain_index(thread_id);
  if (links.length === 0) return violations;

  const by_id = new Map<number, ChainLink>();
  for (const link of links) by_id.set(link.checkpoint_id, link);

  let node: ChainLink | null = links[0] ?? null; // chain_index 按 id 降序
  let walked = 0;
  while (node !== null) {
    walked += 1;
    if (walked > max_walk) {
      violations.push(
        `链遍历超限（>${max_walk} 节点，疑似成环）: 停于 #${node.checkpoint_id}`,
      );
      break;
    }
    const parent =
      node.parent_id !== null && node.parent_id !== undefined
        ? by_id.get(node.parent_id) ?? null
        : null;
    if (node.parent_id !== null && node.parent_id !== undefined && parent === null) {
      const cross = await storage.get_checkpoint(node.parent_id);
      if (cross === null) {
        violations.push(
          `悬挂父指针: #${node.checkpoint_id} -> parent #${node.parent_id} 不存在`,
        );
      } else {
        violations.push(
          `跨线程父指针: #${node.checkpoint_id}(thread=${thread_id}) -> #${cross.checkpoint_id}(thread=${cross.thread_id})`,
        );
      }
      break;
    }
    if (parent !== null) {
      if (parent.checkpoint_id >= node.checkpoint_id) {
        violations.push(
          `父链非递减（环/自引用）: #${node.checkpoint_id} -> #${parent.checkpoint_id}`,
        );
        break;
      }
      if (check_event_seq && parent.event_seq > node.event_seq) {
        violations.push(
          `event_seq 回退: #${node.checkpoint_id} event_seq=${node.event_seq} < 父 #${parent.checkpoint_id} event_seq=${parent.event_seq}`,
        );
      }
    }
    node = parent;
  }
  return violations;
}
